### PR TITLE
Dynamic relocations support

### PR DESCRIPTION
**Pull request by pspcreateprocess**
Adds support of `IMAGE_DYNAMIC_RELOCATION_TABLE` in PE file.
There are also issues about extending format of `IMAGE_LOAD_CONFIG_DIRECTORY` here #220 and here #317.
I added support of load config because it contains dynamic relocation information.

From **winnt.h**:
```
//
// Dynamic value relocation table in loadconfig
//

typedef struct _IMAGE_DYNAMIC_RELOCATION_TABLE {
    DWORD Version;
    DWORD Size;
//  IMAGE_DYNAMIC_RELOCATION DynamicRelocations[0];
} IMAGE_DYNAMIC_RELOCATION_TABLE, *PIMAGE_DYNAMIC_RELOCATION_TABLE;

//
// Dynamic value relocation entries following IMAGE_DYNAMIC_RELOCATION_TABLE
//

#include "pshpack1.h"

typedef struct _IMAGE_DYNAMIC_RELOCATION32 {
    DWORD      Symbol;
    DWORD      BaseRelocSize;
//  IMAGE_BASE_RELOCATION BaseRelocations[0];
} IMAGE_DYNAMIC_RELOCATION32, *PIMAGE_DYNAMIC_RELOCATION32;

typedef struct _IMAGE_DYNAMIC_RELOCATION64 {
    ULONGLONG  Symbol;
    DWORD      BaseRelocSize;
//  IMAGE_BASE_RELOCATION BaseRelocations[0];
} IMAGE_DYNAMIC_RELOCATION64, *PIMAGE_DYNAMIC_RELOCATION64;

typedef struct _IMAGE_DYNAMIC_RELOCATION32_V2 {
    DWORD      HeaderSize;
    DWORD      FixupInfoSize;
    DWORD      Symbol;
    DWORD      SymbolGroup;
    DWORD      Flags;
    // ...     variable length header fields
    // BYTE    FixupInfo[FixupInfoSize]
} IMAGE_DYNAMIC_RELOCATION32_V2, *PIMAGE_DYNAMIC_RELOCATION32_V2;

typedef struct _IMAGE_DYNAMIC_RELOCATION64_V2 {
    DWORD      HeaderSize;
    DWORD      FixupInfoSize;
    ULONGLONG  Symbol;
    DWORD      SymbolGroup;
    DWORD      Flags;
    // ...     variable length header fields
    // BYTE    FixupInfo[FixupInfoSize]
} IMAGE_DYNAMIC_RELOCATION64_V2, *PIMAGE_DYNAMIC_RELOCATION64_V2;

```
Some description of dynamic relocations structures could also be found [here](https://www.alex-ionescu.com/?p=323).
And some example code [here](https://github.com/saferwall/pe/blob/main/loadconfig.go).
Also related to that is a [retpoline presentation](https://www.youtube.com/watch?v=ZfxXjDQRpsU) with details about dynamic relocation types like `IMAGE_IMPORT_CONTROL_TRANSFER_DYNAMIC_RELOCATION`.

### CODE REVIEW

**Comment by erocarrera** [comment]
This is a great PR, thank you!

File: pefile.py [hunk 1]
```diff
@@ -2630,6 +2639,34 @@ class PE:
...
+            "I:12,PageRelativeOffset",
+            "I:1,IndirectCall",
+            "I:19,IATIndex"
+         ),
+    )
+
+    __IMAGE_INDIR_CONTROL_TRANSFER_DYNAMIC_RELOCATION_format__ = (
+        "IMAGE_INDIR_CONTROL_TRANSFER_DYNAMIC_RELOCATION",
+        (
+            "I:12,PageRelativeOffset",
```
Comment (zjgcjy) [inline pefile.py:2654]: type error, which is actually WORD

File: pefile.py [hunk 2]
```diff
@@ -2678,10 +2715,29 @@ class PE:
...
+            "I,GuardLongJumpTargetTable",
+            "I,GuardLongJumpTargetCount",
+            "I,DynamicValueRelocTable",
+            "I,CHPEMetadataPointer",
+            "I,GuardRFFailureRoutine",
+            "I,GuardRFFailureRoutineFunctionPointer",
+            "I,DynamicValueRelocTableOffset",
+            "H,DynamicValueRelocTableSection",
+            "H,Reserved2",
+            "I,GuardRFVerifyStackPointerFunctionPointer"
```
Comment (zjgcjy) [inline pefile.py:2737]: missing comma

**Review by zjgcjy — COMMENTED** [review]
find some mistakes

**Comment by zjgcjy** [comment]
https://github.com/erocarrera/pefile/pull/374 has fixed them

**Pull request merged**